src/app/utils/logic-adaptor.ts

```diff
@@ -0,0 +1,95 @@
+import { cloneDeep, omit } from 'lodash'
+
+import * as Logic from '../../shared/util/logic'
+import {
+  BasicField,
+  FieldResponse,
+  ICheckboxFieldSchema,
+  ICheckboxResponse,
+  IFieldSchema,
+  IFormDocument,
+  IPopulatedForm,
+  IPreventSubmitLogicSchema,
+} from '../../types'
+
+const isCheckboxField = (
+  field: IFieldSchema | undefined,
+): field is ICheckboxFieldSchema => {
+  return field?.fieldType === BasicField.Checkbox
+}
+
+const isCheckboxFieldResponse = (
+  field: FieldResponse,
+): field is ICheckboxResponse => {
+  return field.fieldType === BasicField.Checkbox
+}
+
+const convertServerCheckboxValue = (
+  field: ICheckboxResponse,
+  formFields: IPopulatedForm['form_fields'],
+): Logic.LogicFieldResponse => {
+  const completeField = formFields.find(
+    (formField) => String(formField._id) === String(field._id),
+  )
+  if (!isCheckboxField(completeField)) {
+    // eslint-disable-next-line typesafe/no-throw-sync-func
+    throw new Error(`${completeField} is not a Checkbox field`)
+  }
+
+  const withOthersPrefix = field.answerArray.filter((value) =>
+    value.startsWith('Others: '),
+  )
+  // In-built others option is selected if at least one of them is not
+  // a user-defined option i.e. is one of the field options
+  const inBuiltOthers = withOthersPrefix.find(
+    (value) => !completeField.fieldOptions.includes(value),
+  )
+
+  let others = false
+  if (inBuiltOthers) {
+    field.answerArray.splice(field.answerArray.indexOf(inBuiltOthers), 1)
+    others = true
+  }
+
+  return {
+    ...omit(field, ['answerArray']),
+    answerArray: { options: field.answerArray, others },
+  }
+}
+
+const transformSubmission = (
+  submission: FieldResponse[],
+  form: IFormDocument,
+): Logic.LogicFieldResponse[] => {
+  return submission.map((field) => {
+    const clonedField = cloneDeep(field)
+    if (isCheckboxFieldResponse(clonedField)) {
+      return convertServerCheckboxValue(clonedField, form.form_fields)
+    } else {
+      return clonedField as Logic.LogicFieldResponse
+    }
+  })
+}
+
+export const getVisibleFieldIds = (
+  submission: FieldResponse[],
+  form: IFormDocument,
+): Logic.FieldIdSet => {
+  const transformedSubmission = transformSubmission(submission, form)
+  return Logic.getVisibleFieldIds(transformedSubmission, form)
+}
+
+export const getLogicUnitPreventingSubmit = (
+  submission: FieldResponse[],
+  form: IFormDocument,
+  visibleFieldIds?: Logic.FieldIdSet,
+): IPreventSubmitLogicSchema | undefined => {
+  const transformedSubmission = transformSubmission(submission, form)
+  return Logic.getLogicUnitPreventingSubmit(
+    transformedSubmission,
+    form,
+    visibleFieldIds,
+  )
+}
+
+export * from '../../shared/util/logic'
```

Comment:
think about the principle of encapsulation here - the idea of this adapter is to hide the underlying logic module and only expose what's necessary. is it possible to just export what's necessary instead of freely exposing the underlying module?

Reply:
But I am only re-exposing things that the logic module is currently exposing right? The adaptor should have the same API as the underlying logic module because the adaptor should act as the bridge between the frontend/backend and the logic module. The frontend/backend shouldn't have to access the underlying module at all.

Reply:
>The adaptor should have the same API as the underlying logic module

not sure what you mean by this - isn't the whole point of an adapter to bridge incompatible interfaces by reshaping data?

anyway what's the behaviour if you export multiple functions with the same name from a single module (in this case, `getVisibleFieldIds` and `getLogicUnitPreventingSubmit`?